compiler/rustc_data_structures/src/functor.rs

```diff
@@ -31,6 +35,25 @@ impl<T> IdFunctor for Box<T> {
             raw.assume_init()
         }
     }
+
+    #[inline]
+    fn try_map_id<F, E>(self, mut f: F) -> Result<Self, E>
+    where
+        F: FnMut(Self::Inner) -> Result<Self::Inner, E>,
+    {
+        let raw = Box::into_raw(self);
+        Ok(unsafe {
+            // SAFETY: The raw pointer points to a valid value of type `T`.
+            let value = ptr::read(raw);
+            // SAFETY: Converts `Box<T>` to `Box<MaybeUninit<T>>` which is the
+            // inverse of `Box::assume_init()` and should be safe.
+            let mut raw: Box<mem::MaybeUninit<T>> = Box::from_raw(raw.cast());
+            // SAFETY: Write the mapped value back into the `Box`.
+            ptr::write(raw.as_mut_ptr(), f(value)?);
```

Comment:
So...is this a memory leak on Err?

Reply:
Of what, the box?  It'll be freed when `raw` is dropped (the `MaybeUninit` ensures that the contained value is not dropped—but that's fine, because it was "moved" into `value`).

Reply:
Well, `raw` *after* `f(value)` is written into it. The Box is dropped, but not the value within, right?

Reply:
But `f(value)` will only be written in if it's `Ok` so no early return?